components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -576,16 +610,24 @@ impl VargsRpnFn {
         let fn_ident = &self.item_fn.sig.ident;
         let fn_name = self.item_fn.sig.ident.to_string();
         let arg_type = &self.arg_type;
+        let captures = &self.captures;
         let init_metadata_fn =
             generate_init_metadata_fn(&self.metadata_ctor, &impl_generics, where_clause);
-        let downcast_metadata = if self.metadata_ctor.is_some() {
-            quote! {
-                let metadata = std::any::Any::downcast_ref(metadata).expect("downcast metadata error");
-            }
-        } else {
-            quote! {}
-        };
-        let captures = &self.captures;
+        let downcast_metadata = generate_downcast_metadata(self.metadata_ctor.is_some());
+        let metadata_type_checker = generate_metadata_type_checker(
+            &self.metadata_ctor,
+            &impl_generics,
+            where_clause,
+            |metadata_ctor| {
+                quote! {
+                    let metadata = &#metadata_ctor(expr);
+                    static DUMMY: &[()] = &[];
+                    #fn_ident( #(#captures,)* unsafe {
+                        &*(DUMMY as *const _ as *const [&Option<#arg_type>])
```

Comment:
casting to a raw pointer via reference is considered UB (at least a high-risk operation) whether or not we dereference it, try using `mem::transmute()` instead.

read more: https://github.com/rust-lang/rfcs/pull/2582

Reply:
This function is never called. So I think it's OK.

Reply:
I think the following code is more readable:
```rust
let dummy_args: &'static [&'static Option<#arg_type>]= unsafe { MaybeUninit::uninit().assume_init() };
```

Reply:
Oh, stupid me! `&[]` works perfectly!